FF1Lib/MysterySeedGeneratorLib.cs

```diff
@@ -0,0 +1,1280 @@
+			var freeShip = GetFreeValue(weights.Incentives.FreeItems, FreeItemTypes.Ship, ref freeItems, random);
+			var freeAirship = GetFreeValue(weights.Incentives.FreeItems, FreeItemTypes.Airship, ref freeItems, random);
+			var freeBridge = GetFreeValue(weights.Incentives.FreeItems, FreeItemTypes.Bridge, ref freeItems, random);
+			var freeCanal = GetFreeValue(weights.Incentives.FreeItems, FreeItemTypes.Canal, ref freeItems, random);
+
+
+			// Determine number of allowable loose
+			var guaranteedLocationsCount = weights.Incentives.GuaranteedIncentiveLocations
+				.ConvertAll(ConvertToIncentiveLocationCount).Sum();
+			var guaranteedIncentiveCount = weights.Incentives.GuaranteedIncentiveItems
+				.ConvertAll(ConvertToIncentiveItemCount(freeItems)).Sum();
+			var possibleAdditionalItems = weights.Incentives.IncentiveItemWeights
+				.Where(kvp => kvp.Value > 0)
+				.Select(kvp => ConvertToIncentiveItemCount(freeItems, kvp.Key))
+				.Sum();
+			var possibleAdditionalLocations = weights.Incentives.IncentiveLocationWeights
+				.Where(kvp => kvp.Value > 0)
+				.Select(kvp => ConvertToIncentiveLocationCount(kvp.Key))
+				.Sum();
+
+			var maxLoose = guaranteedIncentiveCount + possibleAdditionalItems - guaranteedLocationsCount;
+			var minLoose = guaranteedIncentiveCount - guaranteedLocationsCount - possibleAdditionalLocations;
+			var maxTrash = guaranteedLocationsCount + possibleAdditionalLocations - guaranteedIncentiveCount;
+			var minTrash = guaranteedLocationsCount - guaranteedIncentiveCount - possibleAdditionalItems;
+
+			var allowedLooseCounts = new HashSet<LooseItemCounts>();
+			if (maxLoose >= 3)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.Loose3);
+			}
+			if (maxLoose >= 2)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.Loose2);
+			}
+			if (maxLoose >= 1)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.Loose1);
+			}
+			if (minLoose <= 0 && 0 <= maxLoose && minTrash <= 0 && maxTrash >= 0)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.None);
+			}
+			if (maxTrash >= 1)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.Trash1);
+			}
+			if (maxTrash >= 2)
+			{
+				allowedLooseCounts.Add(LooseItemCounts.Trash2);
+			}
+
+			var allowedLooseItemWeights = weights.Incentives.LooseItems.Where(x => allowedLooseCounts.Contains(x.Key)).ToDictionary(x => x.Key, x => x.Value);
+			var looseCount = 0;
+			switch (GetEnumFromWeights(allowedLooseItemWeights, random))
+			{
+				case LooseItemCounts.Trash2:
+					looseCount = -2;
+					break;
+				case LooseItemCounts.Trash1:
+					looseCount = -1;
+					break;
+				case LooseItemCounts.None:
+					looseCount = 0;
+					break;
+				case LooseItemCounts.Loose1:
+					looseCount = 1;
+					break;
+				case LooseItemCounts.Loose2:
+					looseCount = 2;
+					break;
+				case LooseItemCounts.Loose3:
+					looseCount = 3;
+					break;
+				default:
+					throw new ArgumentOutOfRangeException();
+			}
+
+			// Determine incentive count and location count
```

Comment:
This is definitely the part that is the most likely to contain bugs (the whole incentive selection part). I tried to test it but it may require unit tests to fully validate.

There is one known issue which is the following:
- You need 10 loose and you have a total of 11 possible items including the ship and canal incentive.
- You select all 9 of the single loose items without selecting the ship and canal.
- It is now "impossible" to get up to 10.

This is not possible with the current weights (also extrememly unlikely in general) but I couldn't think of a good way to fix it while still maintaining the weights